py_ecc/optimized_bls12_381/optimized_swu.py

```diff
@@ -0,0 +1,121 @@
+from typing import (
+    Tuple,
+)
+from py_ecc.fields import (
+    optimized_bls12_381_FQ2 as FQ2,
+)
+from py_ecc.typing import (
+    Optimized_Point3D,
+)
+
+from .constants import (
+    ISO_3_A,
+    ISO_3_B,
+    ISO_3_Z,
+    P_MINUS_9_DIV_16,
+    ETAS,
+    ISO_3_MAP_COEFFICIENTS,
+    POSITIVE_EIGTH_ROOTS_OF_UNITY,
+)
+
+
+# Optimized SWU Map - FQ2 to G2': y^2 = x^3 + 240i * x + 1012 + 1012i
+# Found in Section 4 of https://eprint.iacr.org/2019/403
+def optimized_swu_G2(t: FQ2) -> Tuple[FQ2, FQ2, FQ2]:
+    t2 = t ** 2
+    temp = ISO_3_Z * t2
+    temp = temp + temp ** 2
+    denominator = -(ISO_3_A * temp)  # -a(Z * t^2 + Z^2 * t^4)
+    temp = temp + FQ2.one()
+    numerator = ISO_3_B * temp  # b(Z * t^2 + Z^2 * t^4 + 1)
+
+    # Exceptional case
+    if denominator == FQ2.zero():
+        denominator = ISO_3_Z * ISO_3_A
+
+    # v = D^3
+    v = denominator ** 3
+    # u = N^3 + a * N * D^2 + b* D^3
+    u = (numerator ** 3) + (ISO_3_A * numerator * (denominator ** 2)) + (ISO_3_B * v)
+
+    # Attempt y = sqrt(u / v)
+    (success, sqrt_candidate) = sqrt_division_FQ2(u, v)
+    y = sqrt_candidate
+
+    # Handle case where (u / v) is not square
+    # sqrt_candidate(x1) = sqrt_candidate(x0) * t^3
+    sqrt_candidate = sqrt_candidate * t ** 3
+
+    # u(x1) = Z^3 * t^6 * u(x0)
+    u = (ISO_3_Z * t2) ** 3 * u
+    success_2 = False
+    etas = ETAS
+    for (i, eta) in enumerate(etas):
+        # Valid solution if (eta * sqrt_candidate(x1)) ** 2 * v - u == 0
+        temp1 = eta * sqrt_candidate
+        temp1 = temp1 ** 2 * v - u
+        if temp1 == FQ2.zero() and not success and not success_2:
+            y = sqrt_candidate * eta
+            success_2 = True
+    else:
+        if not success and not success_2:
+            # Unreachable
+            raise Exception("Hash to Curve - Optimized SWU failure")
+
+    if not success:
+        numerator = numerator * t2 * ISO_3_Z
```

Comment:
`iso_3_z_t2` reuse (2)
```suggestion
        numerator = numerator * iso_3_z_t2
```